### ISSUE TITLE
Hyphens not shown when using EB Garamond otf (Android & Linux)

### ISSUE THREAD

**Issue by J-Reis** [koreader/koreader#8290]
* KOReader version: 2021.09
* Device: Samsung Galaxy s9 (also Debian 10.10 on desktop)

#### Issue
When using EB Garamond font, auto-hyphenation works, but no hyphen symbol is inserted at the line break
#### Steps to reproduce

1. load epub
2. select EB Garamond font
3. enable hyphenation (top menu styles, typography rules, hyphenation)

Android log:
[crash.log](https://github.com/koreader/koreader/files/7271140/crash.log)

- The lack of visible hyphens is present with every epub I've tried.
- It persists regardless of what font I choose as for the fallback font.
    - though changing to another main font does resolve the issue
- Changing the hyphenation type (ie algorithmic or dictionary based) makes no difference.

Apologies for over-explaining. But I'll include the following just in case answers relevant questions or is otherwise helpful:

EB Garamond is the only font I'm seeing this issue with and I've attempted to report this to the font developer; but the font does work as expected with many other applications. For instance EB Garamond does show a hyphen when using FBReader with auto-hyphenation on the same device (using the very same otf file in fact). I'm aware that doesn't mean the issue isn't with the font, of course.

The person who filed this (unrelated?) bug report may have been experiencing something similar:
https://github.com/koreader/koreader/issues/7340#issuecomment-786469351

KOReader does also seem to be displaying at least a few non-default letterforms when using EB Garamond. Most obvious is the capital 'W'. The font specimen shows a complex default form as do other programs when using the font (libreoffice, FBReader on Android, Nickel on Kobo Reader, etc.), but KOReader shows an alternate form. Other open type features do seem to be working though as ligatures are shown.

The issue is also present on my Debian laptop using the same version of KOReader installed via .deb
    > this older system is using a different (older) version of EB Garamond as found in the Debian repos, but the behavior is the same.

**Comment by poire-z** [comment]
> select EB Garamond font

You'd have to say which EB Garamond :) because if I pick one I can find on the web, say:
https://bitbucket.org/georgd/eb-garamond/downloads/
I get proper hyphens:
![image](https://user-images.githubusercontent.com/24273478/135711245-e20e18c2-2532-4962-bc91-92672d59077b.png)
(even with no fallback font.)

I'd say the hyphen is broken in _your_ EB Garamond, may be the glyph exists (so it won't be looked for in other fonts) but it is blank with a zero-width (so, you don't see it, and you don't see some blank space taken as in the video in the issue you linked to) ?
Might be worth looking at the font with a font tool like fontforge.
You could try to change Font kerning in the bottom menu from "fast" to "best" to see if Harfbuzz ("best") is able to do better.
May be other software do some additional stuff in some conditions, like if the hyphen is zero-width they pick another glyph? But that's speculation as i don't know what's wrong with your font.
And if it's only one font, and they are alternatives, well, I wouldn't really care :) There are so many broken fonts out there that it's not worth fixing one case of broken font: just pick another font (20% of the fonts I have tried have a too large hyphen, I just trash then when I witness that.)

**Comment by poire-z** [comment]
OK, I get no hyphens with the debian fonts from https://packages.debian.org/bullseye/fonts-ebgaramond
But I have no font tool installed to investigate what's with it.

**Comment by poire-z** [comment]
It seems to have proper info when looking at it with https://fontdrop.info/ :

![image](https://user-images.githubusercontent.com/24273478/135712496-1081ee15-b95d-49b7-b1b4-a2b043938471.png)

but FreeType gives us:
```
10/02/21-12:30:10 DEBUG CreDocument: set font face EB Garamond
hyphen:_slot->metrics.horiAdvance: 0
_slot->metrics.width: 0
 found 0_hyphen_width: 0 index: 109
_hyphen_width: 0 index: 109
_hyphen_width: 0 index: 109
_hyphen_width: 0 index: 109
hyphen:_slot->metrics.horiAdvance: 0
_slot->metrics.width: 0
 found 0_hyphen_width: 0 index: 109
_hyphen_width: 0 index: 109
hyphen:_slot->metrics.horiAdvance: 0
_slot->metrics.width: 0
 found 0_hyphen_width: 0 index: 109
_hyphen_width: 0 index: 109
_hyphen_width: 0 index: 109
```

While with another font:
```
10/02/21-12:27:10 DEBUG CreDocument: set font face EB Garamond Absinthe
hyphen:_slot->metrics.horiAdvance: 640
_slot->metrics.width: 576
 found 10_hyphen_width: 10 index: 112
hyphen:_slot->metrics.horiAdvance: 768
_slot->metrics.width: 640
 found 12_hyphen_width: 12 index: 112
hyphen:_slot->metrics.horiAdvance: 704
_slot->metrics.width: 640
 found 11_hyphen_width: 11 index: 112
hyphen:_slot->metrics.horiAdvance: 704
_slot->metrics.width: 640
 found 11_hyphen_width: 11 index: 112
_hyphen_width: 12
```

So, well, I'm going to say that if it's a bug, it's a FreeType bug :)

**Comment by Frenzie** [comment]
> You'd have to say which EB Garamond :)

Screenshots would also help to illustrate the problem a bit without having to go through 5-10 minutes reproducing just *seeing* it. Which I'm not going to do right now. ;-)

**Comment by J-Reis** [comment]
haha, well first I should thank you for taking the time to investigate something you "wouldn't really care" about! I do appreciate it.

I'm wasn't certain what version my font files were, so I went and got the most recent from https://bitbucket.org/georgd/eb-garamond/downloads/ which appeared to be EBGaramond-0.016.zip. I assume that's the same as you tried. The issue persisted for me even using those newly downloaded files.

2nd paragraph 3rd line "vertigo"
![Screenshot_20211002-053506](https://user-images.githubusercontent.com/23666144/135714579-a82c0694-0d1e-4065-a6f4-aefe96fd9c74.jpg)

to make certain this wasn't a caching issue of some sort I exited KOReader > removed my font files > restarted KOReader > verified it was now using a different font (proving I removed the correct files) > changed the font to freesans > exited the app > extracted the new otf files into my fonts folder > restarted KOReader > changed font to EB Garamond and took the above screenshot.

It appears as though the Debian repo is using the same version (listed as 0.016-4) so it makes sense to me that my two versions on different operating systems don't work, but not why that same version does work for you …

**Comment by poire-z** [comment]
> but not why that same version does work for you …

No, that debian font does NOT work for me: I get the same absence of hyphens as you.
What worked for me was the font downloaded from https://bitbucket.org/georgd/eb-garamond/downloads/.

**Comment by J-Reis** [comment]
Right. Neither the font downloaded from https://bitbucket.org/georgd/eb-garamond/downloads/ or the Debian repo font work for me.

**Comment by J-Reis** [comment]
Right, so whatever the difference is between your installation and mine that makes georgd's 0.016 work for you and not me … your general logic about trying different versions was sound. While earlier georgd versions (I tried 0.015d, 0.015c & 0.014e) seem to behave the same; Mr. Pardo's version here: https://github.com/octaviopardo/EBGaramond12
does show hyphens within KOReader! Indeed it shows the correct capital 'W' form as well … and it appears to function with FBReader, LibreOffice and a few other bits of software just as well as the old version did. So I'm a happy camper.

I still think it is odd, having used this font wherever I could since 2015 that KOReader is the first piece of software to give me any trouble … but I can't argue that some portion of the issue seems to be in the font—since changing it up has fixed it.

Thanks for steering me to a working solution!

**Comment by dcruzbosco** [comment]
I noticed the same problem with EBGaramond08, and switched to a different Garamond. 
But EBGaramond08 gave the perfect book font vibe and I kept coming back to it, using it without hyphenation.

After this discussion, I checked EBGaramond08 in FontForge and found that it is missing the soft-hyphen character.

Here is EBGaramond08 compared with EBGaramond12
![image](https://user-images.githubusercontent.com/1913687/135724849-156c2927-1a6b-478b-8482-09470e0dcfc0.png)
![image](https://user-images.githubusercontent.com/1913687/135725008-f2925b5d-f649-40c2-90e4-c94c1a302b8e.png)


In EBGaramond08, when I copy the hyphen character from the same font to the soft-hyphen and use the modified font, it is showing the hyphens in KOReader.
